src/V0/ArianeeSmartAsset.sol

```diff
@@ -0,0 +1,742 @@
+        ArianeeSmartAssetStorageV0 storage $ = _getArianeeSmartAssetStorageV0();
+        require(
+            block.timestamp < $.idToCertificate[_tokenId].tokenRecoveryTimestamp,
+            "ArianeeSmartAsset: Recovery timestamp reached"
+        );
+
+        address tokenIssuer = $.idToCertificate[_tokenId].tokenIssuer;
+        address tokenOwner = _requireOwned(_tokenId);
+        require(tokenIssuer != tokenOwner, "ArianeeSmartAsset: Issuer is already the owner");
+
+        // _approve(tokenIssuer, _tokenId, address(0)); // TODO: Check if this approve is needed ?
+        _transfer(tokenOwner, tokenIssuer, _tokenId);
+        emit TokenRecovered(_tokenId);
+    }
+
+    /**
+     * @notice Update the status of a recovery request
+     * @dev Must be called by the issuer
+     * @param _tokenId SmartAsset ID
+     * @param _active New status of the recovery request
+     */
+    function updateRecoveryRequest(uint256 _tokenId, bool _active) external isIssuer(_tokenId) whenNotPaused {
+        ArianeeSmartAssetStorageV0 storage $ = _getArianeeSmartAssetStorageV0();
+        $.idToRecoveryRequest[_tokenId] = _active;
+
+        emit RecoveryRequestUpdated(_tokenId, _active);
+    }
+
+    /**
+     * @notice Validate a recovery request and transfer the SmartAsset to the issuer
+     * @dev Can be only called if a recovery request is active and by the `ROLE_ADMIN` of this contract
+     * @param _tokenId SmartAsset ID
+     */
+    function validRecoveryRequest(
+        uint256 _tokenId
+    ) external onlyRole(ROLE_ADMIN) {
+        ArianeeSmartAssetStorageV0 storage $ = _getArianeeSmartAssetStorageV0();
+        require(
+            $.idToRecoveryRequest[_tokenId] == true, "ArianeeSmartAsset: No active recovery request for this SmartAsset"
+        );
+        $.idToRecoveryRequest[_tokenId] = false;
+
+        // _approve(owner(), _tokenId, address(0)); // TODO: Check if this approve is needed ?
+
+        address tokenOwner = _requireOwned(_tokenId);
+        address tokenIssuer = $.idToCertificate[_tokenId].tokenIssuer;
+        _transfer(tokenOwner, tokenIssuer, _tokenId);
+
+        emit RecoveryRequestUpdated(_tokenId, false);
+        emit TokenRecovered(_tokenId);
+    }
+
+    /**
+     * @notice Update the URI of a given SmartAsset
+     * @dev Must be called by the issuer
+     * @param _tokenId SmartAsset ID
+     * @param _uri New URI
+     */
+    function updateTokenURI(uint256 _tokenId, string calldata _uri) external isIssuer(_tokenId) whenNotPaused {
+        _requireOwned(_tokenId);
+
+        ArianeeSmartAssetStorageV0 storage $ = _getArianeeSmartAssetStorageV0();
+        $.idToUri[_tokenId] = _uri;
+
+        emit TokenURIUpdated(_tokenId, _uri);
+    }
+
+    /**
+     * @notice Destroy a given SmartAsset
+     * @dev Must be called by the issuer
+     * @param _tokenId SmartAsset ID
+     */
+    function destroy(
+        uint256 _tokenId
+    ) external whenNotPaused {
```

Comment:
Maybe we should remove this method for now, and add it later since the contract is now upgradable